### PR TITLE
Fix #573

### LINKED ISSUE #573: new -> destroy does not restore preset choices
I expect it restores passed input element to initial state, but was not.

https://github.com/jshjohnson/Choices#destroy
> Usage: Kills the instance of Choices, removes all event listeners and returns passed input to its initial state.

I use the version 7.0.0.
Here is a code: https://gist.github.com/kzkn/43ddb8a52712661c75dfea2d79edeb56

### PR DESCRIPTION
**Pull request by kzkn**
## Description

This PR fixes #573. It contains following fixes:

 - Restores preset choices after `destroy`
 - Loads preset choices only once for each choice after `init`

## How Has This Been Tested?

E2E tests are added.

### CODE REVIEW

File: cypress/integration/select-one.spec.js [hunk 1]
```diff
@@ -861,5 +861,39 @@ describe('Choices - select one', () => {
           });
       });
     });
+
+    describe('new, destroy, init', () => {
```
Comment (jshjohnson) [inline cypress/integration/select-one.spec.js]:
```suggestion
    describe('re-initialising a choices instance', () => {
```

File: cypress/integration/select-one.spec.js [hunk 2]
```diff
@@ -861,5 +861,39 @@ describe('Choices - select one', () => {
...
+      it('has 3 choices', () => {
+        cy.get('[data-test-hook=new-destroy-init]')
+          .find('.choices__list--dropdown .choices__list')
+          .children()
+          .should('have.length', 3)
+      });
+
+      it('has 3 choices after destroy', () => {
```
Comment (jshjohnson) [inline cypress/integration/select-one.spec.js]: This should have a describe around it with a description like `destroying the choices instance` and then the `it` description should be along the lines of `the original select element is preserved`

File: cypress/integration/select-one.spec.js [hunk 3]
```diff
@@ -861,5 +861,39 @@ describe('Choices - select one', () => {
...
+          .find('button.destroy')
+          .click()
+
+        cy.get('[data-test-hook=new-destroy-init]')
+          .find('select')
+          .children()
+          .should('have.length', 3)
+      });
+
+      it('has 3 choices after destroy and init', () => {
```
Comment (jshjohnson) [inline cypress/integration/select-one.spec.js]: See comment above

File: public/test/select-one.html [hunk 4]
```diff
@@ -346,6 +357,14 @@ <h2>Select one inputs</h2>
         new Choices('#choices-set-choice-by-value').setChoiceByValue('Choice 2');
 
         new Choices('#choices-search-by-label', { searchFields: ['label'] });
+
+        const newDestroyInitChoices = new Choices('#choices-new-destroy-init')
+        document.querySelector('button.destroy').addEventListener('click', () => {
+          newDestroyInitChoices.destroy()
+        })
+        document.querySelector('button.init').addEventListener('click', () => {
+          newDestroyInitChoices.init()
```
Comment (jshjohnson) [inline public/test/select-one.html]:
```suggestion
          newDestroyInitChoices.init();
```

File: public/test/select-one.html [hunk 5]
```diff
@@ -346,6 +357,14 @@ <h2>Select one inputs</h2>
         new Choices('#choices-set-choice-by-value').setChoiceByValue('Choice 2');
 
         new Choices('#choices-search-by-label', { searchFields: ['label'] });
+
+        const newDestroyInitChoices = new Choices('#choices-new-destroy-init')
+        document.querySelector('button.destroy').addEventListener('click', () => {
+          newDestroyInitChoices.destroy()
```
Comment (jshjohnson) [inline public/test/select-one.html]:
```suggestion
          newDestroyInitChoices.destroy();
```

File: public/test/select-one.html [hunk 6]
```diff
@@ -346,6 +357,14 @@ <h2>Select one inputs</h2>
...
+        })
+        document.querySelector('button.init').addEventListener('click', () => {
+          newDestroyInitChoices.init()
+        })
```
Comment (jshjohnson) [inline public/test/select-one.html]:
```suggestion
        });
```

File: public/test/select-one.html [hunk 7]
```diff
@@ -346,6 +357,14 @@ <h2>Select one inputs</h2>
         new Choices('#choices-set-choice-by-value').setChoiceByValue('Choice 2');
 
         new Choices('#choices-search-by-label', { searchFields: ['label'] });
+
+        const newDestroyInitChoices = new Choices('#choices-new-destroy-init')
+        document.querySelector('button.destroy').addEventListener('click', () => {
+          newDestroyInitChoices.destroy()
+        })
```
Comment (jshjohnson) [inline public/test/select-one.html]:
```suggestion
        });
```

File: src/scripts/choices.js [hunk 8]
```diff
@@ -219,12 +219,13 @@ class Choices {
...
 
+    this.clearStore();
+
     if (this._isSelectElement) {
       this.passedElement.options = this._presetChoices;
     }
 
-    this.clearStore();
-
+    this._presetChoices = [].concat(this.config.choices);
```
Comment (jshjohnson) [inline src/scripts/choices.js]: Is `this._presetChoices` modified? Not sure why this is needed

**Comment by jshjohnson** [comment]
Thanks for this - i've left some comments 👍

**Comment by perifer** [comment]
Tested this, and can confirm that it works in my app. Great job!

File: src/scripts/choices.js (hunk 8 above)
Comment (kzkn) [reply src/scripts/choices.js]: It was [here](https://github.com/jshjohnson/Choices/blob/0829899284f7a45a9d9ee9cc3902fb14cbfd1bf8/src/scripts/choices.js#L1905-L1917).
I need this change to prepare for the next `init` call. But this change is incomplete.
`this._presetChoices` contains options those are built from `this.config.choices`. They should not restore to `passedElement.options`.

I will fix it with next commit.

Comment (kzkn) [reply src/scripts/choices.js]: I commited 79a4ae0 . It solve the above issue.
Coping `this.config.choices` is necessary. It avoids appending duplicated items on second `init`.

File: cypress/integration/select-one.spec.js (hunk 2 above)
Comment (kzkn) [reply cypress/integration/select-one.spec.js]: Is a0626507edee0a8db4a2d5a16ffdf65df7ac6463 your expected one...?

**Comment by franke-mc** [comment]
@jshjohnson: Maybe you can comment on the current status of this?

**Comment by perifer** [comment]
I'd love to see this resolved. Anything I can do to help?

**Comment by kzkn** [comment]
@jshjohnson re-review please?

File: cypress/integration/select-one.spec.js [hunk 9]
```diff
@@ -861,5 +861,39 @@ describe('Choices - select one', () => {
...
+
+        cy.get('[data-test-hook=new-destroy-init]')
+          .find('.choices__list--dropdown .choices__list')
+          .children()
+          .should('have.length', 3);
+      });
+    });
+
+    describe('destroying the choices instance', () => {
+      it('the original select element is preserved', () => {
```
Comment (jshjohnson) [inline cypress/integration/select-one.spec.js]:
```suggestion
      it('preserves the original select element', () => {
```

File: cypress/integration/select-one.spec.js [hunk 10]
```diff
@@ -861,5 +861,39 @@ describe('Choices - select one', () => {
           });
       });
     });
+
+    describe('re-initialising a choices instance', () => {
+      it('the original choices element is preserved', () => {
```
Comment (jshjohnson) [inline cypress/integration/select-one.spec.js]:
```suggestion
      it('preserves the choices list', () => {
```

File: public/test/select-one.html [hunk 11]
```diff
@@ -346,6 +357,14 @@ <h2>Select one inputs</h2>
         new Choices('#choices-set-choice-by-value').setChoiceByValue('Choice 2');
 
         new Choices('#choices-search-by-label', { searchFields: ['label'] });
+
+        const newDestroyInitChoices = new Choices('#choices-new-destroy-init')
```
Comment (jshjohnson) [inline public/test/select-one.html]:
```suggestion
        const newDestroyInitChoices = new Choices('#choices-new-destroy-init');
```

**Comment by kzkn** [comment]
@jshjohnson
I had fixed new indicated items from you and a conflict with master branch.

Then, I have gotten a new ambiguity.
Should the `Choices` selection be reproduced in the original DOM after calling `destroy`?

```html
<select>
<option selected>one</option>
<option>two</option>
<option>three</option>
</select>

<script>
const select = document.querySelector('select')
console.log(select.value) // => one
const choices = new Choices(select)
// select "two" on UI
choices.destroy()
console.log(select.value) // what?
</script>
```

How do you think it should work?

Thanks!

**Comment by jshjohnson** [comment]
> Should the `Choices` selection be reproduced in the original DOM after calling `destroy`?

For a first pass I would say no. We can revisit this is people want to though

**Comment by kzkn** [comment]
@jshjohnson 
> For a first pass I would say no. We can revisit this is people want to though.

Thanks! I got it.

> P.s. you have some conflicts!

I have fixed.

**Comment by filip-jk** [comment]
Can't wait for this to be released!

File: src/scripts/choices.js [hunk 12]
```diff
@@ -144,6 +144,7 @@ class Choices {
         classNames: this.config.classNames,
         template: data => this._templates.option(data),
       });
+      this._presetOptions = this.passedElement.options;
```
Comment (jshjohnson) [inline src/scripts/choices.js]: Could this be moved above line 186 please? That's where all the "preset" variables are defined

Comment (kzkn) [reply src/scripts/choices.js]: Thanks! I have fixed.

**Comment by filip-jk** [comment]
@kzkn Thank you, Sir!

@jshjohnson Thank you, as well! Is there any chance to get this updated version on NPM? :)

**Comment by jshjohnson** [comment]
@filip-jk this has been released with 9.0.1 🎉 thanks @kzkn

**Comment by filip-jk** [comment]
@jshjohnson Ace, thanks!

**Comment by filip-jk** [comment]
@jshjohnson Sorry for bugging you about this but I think there is an issue with the destroy() method. It doesn't seem to pass the selected value? 

I'm on 9.0.1 and let's say, the select I initialize Choices on has the selected value of "3". I destroy that Choices object and, unfortunately, the selected value on the <select> tag is not "3". It gets "reset" I believe.

**Comment by jshjohnson** [comment]
That's expected @filip-jk https://github.com/jshjohnson/Choices/pull/574#issuecomment-553794506

**Comment by filip-jk** [comment]
@jshjohnson I'm confused. I'm initializing Choices on a select that has a selected value of "5". Choices take that selected value and I can see it in the rendered Choices dropdown menu. Once I destroy it, the select gets totally reset so it's not really the states the Choices was initialized on, right?

If that's expected than the description of the method is wrong:

"Kills the instance of Choices, removes all event listeners and returns passed input to its initial state."

It's not the initial state if the selected option/value is different.

**Comment by jshjohnson** [comment]
Yeah you're right, I would expect the originally selected option to persist. I'll investigate

**Comment by tagliala** [comment]
Hi, is it possible to reopen this issue? I'm experiencing the same issue described by @filip-jk

The use case is in a stimulus.js + turbo application, I will try to provide a reduced test case, but I'm going to describe it

1. Open a page with a Choices.js input with a selected value of 3
2. Go back in history (that will call `destroy` on the Choices.js instance)
3. Navigate forward to the page of Choices.js (that will trigger a `init` on the Choices.js instance)
4. I expect `3` to be selected, but the selection is lost

**Pull request merged**